### DISCUSSION TITLE
Question: Publishing apps

### DISCUSSION THREAD

**Discussion by deepa** [nrwl/nx#10762] [General]
I'm somewhat new to Nx and wanted to understand more about the differences between Nx applications and libraries. I see that libraries can be publishable. Can apps be published and consumed too? If yes, should it be done similar to publishing and consuming a lib (building the app, adding a package.json manually and publishing it from the dist directory)? Or, is it not recommended to have publishable code in apps and have it in libs? Thanks!

**Comment by rfprod** [comment]
@deepa I can share my vision, maybe it will help.

Based on my personal mental model, I rather view applications as deployables. At the same time, an application can be a CLI, which means it is delivered as a downloadable/installable package rather than a deployable. In such case maybe it makes sense to publish it somewere for the users.
Additionally, again based on my understanding of Nx which may be incorrect, an application is something that nothing else depends on in terms of source code, i.e. a library source code can depend on a library source code, an application source code can depend on a library source code, but an application source code **cannot** depend on an application source code, a library source code **cannot** depend on an application source code.

So, lets say one wants to create ~~an application~~ a program (I'll use this term for comprehensiveness) that can be interacted with using a CLI, but at the same time this program source code should be used by a deployable backend application that resides in the `apps` workspace folder. In such case, the source code of the program goes to the `libs` directory.

If one were to create a program that works only as a CLI, I would recommend to put its source code into the `libs` directory as well, just to keep it extensible/reusable at some point in the future.

And in general, maybe my perception is not correct, but an application is something like an end product that serves a very specific narrow purpose. Why would someone publish an application?

**Reply by kwiniarski97** [reply]
This would be useful when e.g. your apps is deployed as part of another application. e.g. I might have an application that would be deployed as part of .NET server in an iframe. In this case, it's easiest when you publish it as a package and include it in the .NET server with proper version.
